virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+        """
+
+        self.ppfd = calculate_ppfd(data["shortwave_in"], data["elevation"])
+        self.topofcanopy_radiation = calculate_topofcanopy_radiation(
+            data["shortwave_in"], data["elevation"]
+        )
+        self.longwave_radiation = calculate_longwave_radiation(
+            data["canopy_temperature"], data["surface_temperature"]
+        )
+        self.netradiation_surface = calculate_netradiation_surface(
+            self.topofcanopy_radiation,
+            data["canopy_absorption"],
+            self.longwave_radiation,
+        )
+
+
+# helper functions
+def calculate_atmospheric_transmissivity(
+    elevation: NDArray[np.float32],
+    sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+    cloudy_transmissivity: float = CLOUDY_TRANSMISSIVITY,
+    transmissivity_coefficient: float = TRANSMISSIVITY_COEFFICIENT,
+    beer_regression: float = BEER_REGRESSION,
+) -> NDArray[np.float32]:
+    """Calculate atmospheric transmissivity (tau).
+
+    Args:
+        elevation: elevation above sea level, [m]
+        sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+            and 1 (cloud free sky), default = 1
+        cloudy_transmissivity: cloudy transmittivity :cite:p:`Linacre1968`, default set
+            in config
+        transmissivity_coefficient: angular coefficient of transmittivity
+            :cite:p:`Linacre1968`, default set in config
+        beer_regression: parameter in equation for atmospheric transmissivity based on
+            regression of Beer's radiation extinction function :cite:p:`Allen1996`,
+            default set in config
+
+    Returns:
+        atmospheric transmissivity, unitless
+    """
+
+    # check sunshine fraction between 0 and 1
+    if 0 > np.any(sunshine_fraction) > 1:
+        to_raise = ValueError(
+            "The fraction of sunshine hours needs to be between 0 and 1!"
+        )
+        LOGGER.critical(to_raise)
+        raise to_raise
+
+    # calculate transmissivity (tau), unitless
+    tau_o = cloudy_transmissivity + transmissivity_coefficient * sunshine_fraction
+    return tau_o * (1.0 + beer_regression * elevation)
+
+
+def calculate_ppfd(
+    shortwave_in: NDArray[np.float32],
+    elevation: NDArray[np.float32],
+    sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+    albedo_vis: NDArray[np.float32] = np.array(0.03, dtype=np.float32),
+    flux_to_energy: float = FLUX_TO_ENERGY,
+) -> NDArray[np.float32]:
+    """Calculate top of canopy photosynthetic photon flux density, [mol m-2].
+
+    Args:
+        shortwave_in: downward shortwave radiation, [J m-2]
+        elevation: elevation above sea level, [m]
+        sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+            and 1 (cloud free sky), default = 1
+        albedo_vis: visible light albedo, default = 0.03
+        flux_to_energy: flux to energy conversion factor, [umol J-1],default from config
```

Comment:
Out of interest is umol/J the usual unit for this? If so that's ok, but if there's not a standard way of representing it, I'd just use SI units.

Reply:
According to the reference, for daily values, the unit is moles per megajoules, but the hourly values were expressed in micromoles per joule to maintain consistency with energy units of global solar radiation energy-based photosynthetically active radiation (PARE) at hourly and daily scales. I guess I could introduce a scaling factor if I wanted t express the constant in mol J-1.